### PR TITLE
Policy Cookbook and Library Examples (main)

### PR DESCRIPTION
**Pull request by korydraughn**
- Introduces a new section for capturing known policy techniques to various problems
- Absorbs the https://github.com/irods/irods_library_examples

### CODE REVIEW

File: docs/developers/library_examples.md
Comment (trel) [inline docs/developers/library_examples.md:5]: still true?

File: docs/policy_cookbook/metadata_based_synchronization.md
Comment (trel) [inline docs/policy_cookbook/metadata_based_synchronization.md:1]: this title does not match the filename - so one or the other should probably get updated.

File: mkdocs.yml
Comment (trel) [inline mkdocs.yml:73]: not sure about 'Home'... but don't have a better suggestion at the moment...

**Review by trel — COMMENTED** [review]
does this render the left sidebar correctly now?

**Comment by korydraughn** [comment]
> does this render the left sidebar correctly now?

Yes

File: docs/policy_cookbook/metadata_based_synchronization.md
Comment (korydraughn) [reply docs/policy_cookbook/metadata_based_synchronization.md:1]: Will correct that. The filename is the correct name.

Still a WIP

File: mkdocs.yml
Comment (korydraughn) [reply mkdocs.yml:73]: Same here

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:5]: Maybe. I'm still thinking about this one.

Doxygen may be enough for the C APIs. However, it does seem like a good idea to provide examples around common operations. Some users will be afraid to use C++ due to complexity. For that reason, C may be a better fit. Here are a few things I feel would be good to have here for everyone, including the core dev team:
- Connecting to iRODS
- Iterating over a collection
- Using GenQuery

The Library Examples section can serve as a quick reference for various things instead of requiring devs to look through Doxygen documentation.

Comment (trel) [reply docs/developers/library_examples.md:5]: reasonable list.   i think that makes sense.

File: mkdocs.yml
Comment (alanking) [reply mkdocs.yml:73]: Here are some suggestions, which may or may not be better...
 - Index
 - Table of Contents
 - Overview
 - Introduction

Alternatively, this could all go under one page (much like the library examples or troubleshooting pages) and stuck under System Overview or even grow another section for administrators.

File: docs/policy_cookbook/index.md
Comment (alanking) [inline docs/policy_cookbook/index.md:5]: Overtime -> "Over time"

File: docs/developers/library_examples.md
Comment (alanking) [inline docs/developers/library_examples.md:11]: Any advantage to linking to Doxygen pages from here? Is that even possible?

**Review by alanking — COMMENTED** [review]
Very exciting

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:11]: Sounds like a good idea to me.

Comment (trel) [reply docs/developers/library_examples.md:11]: yes, you can link to it, same as in the sidebar...    if you need something more targeted, then I'm not sure how stable the generated URLs are over time/versions.

File: docs/policy_cookbook/metadata_based_synchronization.md
Comment (korydraughn) [reply docs/policy_cookbook/metadata_based_synchronization.md]: Fixed title

File: docs/policy_cookbook/index.md
Comment (korydraughn) [reply docs/policy_cookbook/index.md]: fixed

File: docs/policy_cookbook/metadata_based_synchronization.md
Comment (trel) [reply docs/policy_cookbook/metadata_based_synchronization.md]: filename itself?

**Comment by korydraughn** [comment]
> I’m not familiar with the second parameter…

`msiSleep('seconds', 'microseconds')`

File: mkdocs.yml
Comment (korydraughn) [reply mkdocs.yml]: Updated navigation.

Added section called `Administrators` and placed the `Policy Cookbook` under that.

`System Overview` could work, but the word, `Overview`, doesn't fit the information IMO.

Thoughts?

File: docs/developers/library_examples.md
Comment (korydraughn) [inline docs/developers/library_examples.md:960]: These can be individual issues.

**Comment by korydraughn** [comment]
The information in this files is probably a good starting point.

I need to either remove the http links or make sure they stay up to date without us having to watch them constantly.

File: docs/developers/library_examples.md
Comment (trel) [inline docs/developers/library_examples.md:590]: now 'main' branch?

File: docs/developers/library_examples.md
Comment (trel) [inline docs/developers/library_examples.md:619]: nothing on right hand side?

File: docs/developers/library_examples.md
Comment (trel) [inline docs/developers/library_examples.md:728]: same

File: docs/developers/library_examples.md
Comment (trel) [inline docs/developers/library_examples.md:73]: one of 8 in this page - no right hand side?

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:922]: These URLs may be cut from the final version.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:619]: That is correct. The examples are not meant to be runnable as they are. For some things, it is on the reader to fill in gaps.

The library examples also contain examples for connecting to an iRODS server so it shouldn't be difficult for the reader to add that.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: removed

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: removed

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:358]: I'll add a link to the doxygen documentation at the top of the page. Like @trel said, it may be troublesome to manage links if they change between versions. We can also revisit this later.

File: docs/developers/library_examples.md
Comment (alanking) [inline docs/developers/library_examples.md:703]: As a best practice, maybe we should use `at` here? Or mention it?

File: docs/developers/library_examples.md
Comment (alanking) [inline docs/developers/library_examples.md:721]: So smol

File: docs/developers/library_examples.md
Comment (alanking) [reply docs/developers/library_examples.md:632]: Maybe we can have `conn` be passed in like in the `print_all_resource_names` example under `query`?

**Review by alanking — COMMENTED** [review]
I think having everything on two pages is much better, personally. Overall this is amazing, my suggestions are nice-to-haves

Should we bring this out of draft?

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:632]: Very good idea. Will change.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:15]: That's correct.

Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:15]: It would be great if we could create a color scheme for the rule language.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:703]: Absolutely.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:181]: I was trying to emphasize that those comments are specific to the parameter. Do you feel that isn't conveyed well enough?

Comment (alanking) [reply docs/developers/library_examples.md:181]: I see what you mean. I can't think of a *better* way, but an alternative could be to place the comment off to the side or above and reference the 3rd input in each instance, it just kind of looked like a mistake to me when I was reading through. Maybe it's clear to others as it is.

Comment (korydraughn) [reply docs/developers/library_examples.md:181]: I'm looking at the rendered version and I agree. I'll change it.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md]: Done.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md]: Done.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md]: Done.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:195]: I'm not sure what that blurb would say. Do you have words that you feel should be mentioned here?

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:359]: True. I've updated the text to link against things that are stable.

File: mkdocs.yml
Comment (korydraughn) [reply mkdocs.yml]: @trel @alanking Any thoughts on the navigation?

You can see the current rendering by going to kdd-ws.euro...org:8080

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md:968]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/developers/library_examples.md
Comment (korydraughn) [reply docs/developers/library_examples.md]: Done.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md]: Done.

File: docs/administrators/policy_cookbook.md
Comment (trel) [reply docs/administrators/policy_cookbook.md:195]: how's this...

> The original RescQuota system was geared to count bytes on storage resources... either per user or group for a single resource, or per user or group across all resources.  Since 4.3.0, user quotas have been removed (user and group quotas could not be enforced simultaneously - see irods/irods#4481)- but group quotas can still be set and enforced across all resources.  In this case, we're interested in setting a quota for a group which contains a single user (equivalent to setting a user quota).

Maybe we should start referring to this quota system as "Physical Quotas"...  wondering out loud if we can pull them out into a separate rule engine plugin...   gah, that sounds like a lot of work.

Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:195]: > The original RescQuota system was geared to count bytes on storage resources... either per user or group for a single resource, or per user or group across all resources. Since 4.3.0, user quotas have been removed (user and group quotas could not be enforced simultaneously - see irods/irods#4481)- but group quotas can still be set and enforced across all resources. In this case, we're interested in setting a quota for a group which contains a single user (equivalent to setting a user quota).

My gut reaction is that feels like something that should be linked to rather than placed directly in the cookbook. Although, technical books tend to have little pockets of text like that at various places.

I'm going to add it to see what it's like in the rendered version.

> Maybe we should start referring to this quota system as "Physical Quotas"... wondering out loud if we can pull them out into a separate rule engine plugin... gah, that sounds like a lot of work.

Why "Physical Quotas"? I feel we can externalize them into a REP. Unlike Logical Quotas, this system does not react to data movement in real time, so it may be simpler to implement. Removing the code from the server is probably where the challenge lies.

Comment (trel) [reply docs/administrators/policy_cookbook.md:195]: That's fair about living somewhere else and linking... but, not sure where that would be yet.

--

Physical Quotas, rather than Logical Quotas.   Agreed that removing the `iadmin` subcommands would be a hairy process. And providing a(n automatic) transition/upgrade process is not clear either...

Comment (alanking) [reply docs/administrators/policy_cookbook.md:195]: I mean, I just legitimately didn't know why we had to turn on resc quotas to enable this feature. :)

A single sentence like: "For historical reasons, this policy enables tracking quota usage." Or something like that would be fine by me. Fleshing that out more as @trel did could be valuable too, of course.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:15]: Marking as resolved. We can always change the language used for code samples.

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:195]: Sure. I agree that @trel's statement needs to be added to the documentation. I may look into adding that somewhere via this PR.

Is this about protecting the reader from confusion due to the name of the microservice and NREP rule?

File: docs/administrators/policy_cookbook.md
Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:138]: Well, this info is available in the post PEP because of the `*DATA_OBJ_INPUT`.

I'll make sure to mention that the pre PEP always fires before the post PEP. I will also include a link to the dynamic PEPs page.

File: docs/administrators/policy_cookbook.md
Comment (trel) [reply docs/administrators/policy_cookbook.md:200]: yes, the Resc in the name.... when we're attempting to set a 'user' quota.

Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:200]: I thought so. Take a look at the current write-up and let me know what you think. It is closer to what @alanking proposed.

Comment (alanking) [reply docs/administrators/policy_cookbook.md:200]: I think it is sufficient, yep

Comment (trel) [reply docs/administrators/policy_cookbook.md:200]: Agreed

Comment (korydraughn) [reply docs/administrators/policy_cookbook.md:200]: Cool. Marking as resolved.

File: mkdocs.yml
Comment (trel) [reply mkdocs.yml]: i like the navigation so far.   I think it's in the right place for now.

Comment (korydraughn) [reply mkdocs.yml]: Excellent.

**Comment by korydraughn** [comment]
I think this is in a good spot to be merged.

If there aren't any objections, then I'll go ahead and add the #'s.

**Comment by trel** [comment]
agreed.

**Comment by korydraughn** [comment]
Added #'s

**Comment by korydraughn** [comment]
Don't merge yet. I need to update one thing.

**Comment by korydraughn** [comment]
It's ready to be merged now.

Had to update the sync delay rules example.

**Review by alanking — APPROVED** [review]

**Pull request merged**
